libvast/test/pattern.cpp

```diff
@@ -14,94 +14,44 @@
+}
+} // namespace
+
 TEST(functionality) {
   std::string str = "1";
-  CHECK(pattern("[0-9]").match(str));
-  CHECK(!pattern("[^1]").match(str));
+  CHECK(make_pattern("[0-9]").match(str));
+  CHECK(!make_pattern("[^1]").match(str));
   str = "foobarbaz";
-  CHECK(pattern("bar").search(str));
-  CHECK(!pattern("bar").search("FOOBARBAZ"));
-  CHECK(!pattern("^bar$").search(str));
-  CHECK(pattern("^\\w{3}\\w{3}\\w{3}$").match(str));
-  CHECK(pattern::glob("foo*baz").match(str));
-  CHECK(pattern::glob("foo???baz").match(str));
-  CHECK(pattern::glob(str).match(str));
+  CHECK(make_pattern("bar").search(str));
+  CHECK(!make_pattern("bar").search("FOOBARBAZ"));
+  CHECK(!make_pattern("^bar$").search(str));
+  CHECK(make_pattern("^\\w{3}\\w{3}\\w{3}$").match(str));
   str = "Holla die Waldfee!";
-  pattern p{"\\w+ die Waldfe{2}."};
+  auto p = make_pattern("\\w+ die Waldfe{2}.");
   CHECK(p.match(str));
   CHECK(p.search(str));
-  p = pattern("(\\w+ )");
+  p = make_pattern("(\\w+ )");
   CHECK(!p.match(str));
   CHECK(p.search(str));
 }
 
 TEST(comparison with string) {
-  auto rx = pattern{"foo.*baz"};
+  auto rx = make_pattern("foo.*baz");
   CHECK("foobarbaz"sv == rx);
   CHECK(rx == "foobarbaz"sv);
 }
 
-TEST(composition) {
-  auto foo = pattern{"foo"};
-  auto bar = pattern{"bar"};
-  auto foobar = "^" + foo + bar + "$";
-  CHECK(foobar.match("foobar"));
-  CHECK(!foobar.match("FOOBAR"));
-  CHECK(!foobar.match("fooBAR"));
-  CHECK(!foobar.match("FOObar"));
-  CHECK(!foobar.match("foo"));
-  CHECK(!foobar.match("bar"));
-  auto foo_or_bar = foo | bar;
-  CHECK(!foo_or_bar.match("foobar"));
-  CHECK(foo_or_bar.search("foobar"));
-  CHECK(foo_or_bar.match("foo"));
-  CHECK(foo_or_bar.match("bar"));
-  CHECK(!foo_or_bar.match("FOO"));
-  CHECK(!foo_or_bar.match("BAR"));
-  auto foo_and_bar = foo & bar;
-  CHECK(foo_and_bar.search("foobar"));
-  CHECK(foo_and_bar.match("foobar"));
-  CHECK(!foo_and_bar.search("FOOBAR"));
-  CHECK(!foo_and_bar.match("FOOBAR"));
-  CHECK(!foo_and_bar.match("foo"));
-  CHECK(!foo_and_bar.match("bar"));
-}
-
-TEST(composition with case insensitivity) {
-  auto foo = pattern{"foo", true};
-  auto bar = pattern{"bar"};
-  auto comp = foo + bar;
-  CHECK(comp.case_insensitive());
-  comp = bar + foo;
-  CHECK(!comp.case_insensitive());
-  comp = foo | bar;
-  CHECK(comp.case_insensitive());
-  comp = bar | foo;
-  CHECK(!comp.case_insensitive());
-  comp = foo & bar;
-  CHECK(comp.case_insensitive());
-  comp = bar & foo;
-  CHECK(!comp.case_insensitive());
-  comp = bar + "str";
-  CHECK(!comp.case_insensitive());
-  comp = "str" + bar;
-  CHECK(!comp.case_insensitive());
-  comp = foo + "str";
-  CHECK(comp.case_insensitive());
-  comp = "str" + foo;
-  CHECK(comp.case_insensitive());
-  foo += bar;
-  CHECK(foo.case_insensitive());
-  bar += foo;
-  CHECK(!bar.case_insensitive());
-}
-
 TEST(case insensitive) {
-  auto pat = pattern("bar", true);
+  auto pat = unbox(pattern::make("bar", true));
```

Comment:
This could be named bool. Hard to tell what `true` means when reading the code.

Reply:
It's also a tiny bit confusing to see `make_pattern` everywhere else in these tests.

Reply:
I added an optional `case_insensitive` parameter to the `make_pattern` test helper function in https://github.com/tenzir/vast/pull/2964/commits/44b5a18cd6f4c783564dfffa4c636c5b3d9bf21c

Reply:
My first comment was trying to say: perhaps use an `enum` instead of `bool` with a crisp name? IIRC we do that for almost all booleans in public interfaces nowadays.

Reply:
So basically replace the boolean `pattern::make` argument with a `case_insensitivity` enum?

Reply:
I would call the enum perhaps `pattern_flags` with a `default` and `case_insensitive` mode?

Reply:
As we are planning to extend this, we probably will eventually make these flags _additive_, so that you can have things like `/.../asdfi`.

Reply:
Then maybe it would be better to create a struct `pattern_flags` that contains booleans?

Reply:
Yeah, good idea. But then let's call it `pattern_options` for maximum flexibility.

Reply:
I implemented a `pattern::options` struct in https://github.com/tenzir/vast/pull/2964/commits/377f12a524af2dff5c5506a2a80f4bfe7deeff24